dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -26,83 +26,52 @@ namespace
 void checkPacketSize(size_t size)
 {
     static constexpr size_t max_packet_size = 1u << 31;
-    if (size >= max_packet_size)
+    if (unlikely(size >= max_packet_size))
         throw Exception(fmt::format("Packet is too large to send, size : {}", size));
 }
 
+TrackedMppDataPacketPtr serializePacket(tipb::SelectResponse & response)
+{
+    auto tracked_packet = std::make_shared<TrackedMppDataPacket>();
+    tracked_packet->serializeByResponse(response);
+    checkPacketSize(tracked_packet->getPacket().ByteSizeLong());
+    return tracked_packet;
+}
 } // namespace
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::clearExecutionSummaries(tipb::SelectResponse & response)
+void MPPTunnelSetBase<Tunnel>::sendExecutionSummary(tipb::SelectResponse & response)
 {
-    /// can not use response.clear_execution_summaries() because
-    /// TiDB assume all the executor should return execution summary
-    for (int i = 0; i < response.execution_summaries_size(); i++)
-    {
-        auto * mutable_execution_summary = response.mutable_execution_summaries(i);
-        mutable_execution_summary->set_num_produced_rows(0);
-        mutable_execution_summary->set_num_iterations(0);
-        mutable_execution_summary->set_concurrency(0);
-    }
+    RUNTIME_CHECK(!tunnels.empty());
+    // for execution summary, only need to send to one tunnel.
+    tunnels[0]->write(serializePacket(response));
 }
 
 template <typename Tunnel>
 void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
 {
-    TrackedMppDataPacket tracked_packet;
-    tracked_packet.serializeByResponse(response);
-    tunnels[0]->write(tracked_packet.getPacket());
-
-    if (tunnels.size() > 1)
-    {
-        /// only the last response has execution_summaries
-        if (response.execution_summaries_size() > 0)
-        {
-            clearExecutionSummaries(response);
-            tracked_packet = TrackedMppDataPacket();
-            tracked_packet.serializeByResponse(response);
-        }
-        for (size_t i = 1; i < tunnels.size(); ++i)
-            tunnels[i]->write(tracked_packet.getPacket());
-    }
+    // for root mpp task, only one tunnel will connect to tidb/tispark.
```

Comment:
The comment is out of date, pass through write now can also exists between two TiFlash nodes.

Reply:
The pass through write between tiflash is executed by `broadcastOrPassThroughWrite` now.
`MPPTunnelSet::write` only used for root mpp task.